### PR TITLE
Feature/magnetometer

### PR DESCRIPTION
**Pull request by pvechersky**
Adding a macro for a magnetometer component and a plugin to publish magnetic field data on a ROS topic

### CODE REVIEW

**Comment by pvechersky** [comment]
Pull request initially failed the checks on the build server due to use of Gazebo API version newer than 5. Now I made a magnetometer plugin that is independent of Gazebo sensor model.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_magnetometer_plugin.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2015 Fadri Furrer, ASL, ETH Zurich, Switzerland
```
Comment (ffurrer) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_magnetometer_plugin.h]: You should add yourself here as an author instead of us ;)

**Comment by pvechersky** [comment]
Thank you, @ffurrer! :) ... Should I add myself 'instead' of you guys or append my name to the list?

File: rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+  // Initialize the reference magnetic field vector in world frame, taking into
+  // account the initial bias
+  mag_W_ = math::Vector3(ref_mag_field.X() + initial_bias[0](random_generator_),
+                         ref_mag_field.Y() + initial_bias[1](random_generator_),
+                         ref_mag_field.Z() + initial_bias[2](random_generator_));
+
+  // Fill the magnetometer message
+  mag_message_.header.frame_id = frame_id_;
+  mag_message_.magnetic_field.x = mag_W_.x;
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp]: why is this needed? it get's overwritten anyway before publishing

File: rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  mag_message_.magnetic_field.y = mag_W_.y;
+  mag_message_.magnetic_field.z = mag_W_.z;
+  mag_message_.magnetic_field_covariance[0] = noise_normal.X() * noise_normal.X();
+  mag_message_.magnetic_field_covariance[4] = noise_normal.Y() * noise_normal.Y();
+  mag_message_.magnetic_field_covariance[8] = noise_normal.Z() * noise_normal.Z();
+}
+
+void GazeboMagnetometerPlugin::OnUpdate(const common::UpdateInfo& _info) {
+  // Get the current pose and time from Gazebo
+  math::Pose world_pose = link_->GetWorldPose();
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp]: `math::Pose T_W_B =`
We use the convention here: https://github.com/ethz-asl/minkindr/wiki/Common-Transformation-Conventions

File: rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  common::Time current_time  = world_->GetSimTime();
+
+  // Calculate the magnetic field noise.
+  math::Vector3 mag_noise(noise_n_[0](random_generator_),
+                          noise_n_[1](random_generator_),
+                          noise_n_[2](random_generator_));
+
+  // Rotate the earth magnetic field into the inertial frame
+  math::Vector3 field = world_pose.rot.RotateVectorReverse(mag_W_ + mag_noise);
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp]: field_B

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_magnetometer_plugin.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#include <sensor_msgs/MagneticField.h>
+
+#include "rotors_gazebo_plugins/common.h"
+#include "rotors_gazebo_plugins/sdf_api_wrapper.hpp"
+
+namespace gazebo {
+// Default magnetic field [Tesla] in NED frame, obtained from World Magnetic
+// Model: (http://www.ngdc.noaa.gov/geomag-web/#igrfwmm) for Zurich:
+// lat=+47.3667degN, lon=+8.5500degE, h=+500m, WGS84
+static const SdfVector3 kDefaultRefMagField(0.000021493, 0.000000815, 0.000042795);
```
Comment (burrimi) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_magnetometer_plugin.h]: `static constexpr SdfVector3`

**Comment by burrimi** [comment]
thanks a lot @pvechersky. scolgtm.
